### PR TITLE
#3210 Fix for "Texture will be downscaled" happening too often.

### PR DESCRIPTION
**Pull request by RunitaiLinden**

### CODE REVIEW

File: indra/newview/llviewertexture.cpp [hunk 1]
```diff
@@ -559,7 +559,10 @@ void LLViewerTexture::updateClass()
         // lower discard bias over time when free memory is available
         if (sDesiredDiscardBias > 1.f && over_pct < 0.f)
         {
-            sDesiredDiscardBias -= gFrameIntervalSeconds * 0.01f;
+            static LLCachedControl<F32> high_mem_discard_increment(gSavedSettings, "RenderHighMemMinDiscardDecrement", .1f);
```
Comment (akleshchev) [inline indra/newview/llviewertexture.cpp:562]: Typo:
high_mem_discard_**increment**

Also won't it be better to make this dynamic? The more memory is free the faster bias falls?

Comment (RunitaiLinden) [reply indra/newview/llviewertexture.cpp:562]: See line below, it is faster when more memory is free (over_pct is negative), fixing typo...

File: indra/newview/llviewertexturelist.cpp [hunk 2]
```diff
@@ -1066,13 +1066,26 @@ F32 LLViewerTextureList::updateImagesCreateTextures(F32 max_time)
...
+
+        // desired discard may change while an image is being decoded. If the texture in VRAM is sufficient
+        // for the current desired discard level, skip the texture creation.  This happens more often than it probably
+        // should
+        bool redundant_load = imagep->hasGLTexture() && imagep->getDiscardLevel() <= imagep->getDesiredDiscardLevel();
+
+        if (!redundant_load)
+        {
+           imagep->createTexture();
+        }
```
Comment (akleshchev) [inline indra/newview/llviewertexturelist.cpp:1078]: May be add else LL_DEBUGS() here?

File: indra/newview/llviewertexture.cpp (hunk 1 above)
Comment (akleshchev) [reply indra/newview/llviewertexture.cpp]: Sorry, somehow missed that.

File: indra/newview/llviewertexturelist.cpp [hunk 3]
```diff
@@ -1066,13 +1066,26 @@ F32 LLViewerTextureList::updateImagesCreateTextures(F32 max_time)
...
+        
         imagep->postCreateTexture();
         imagep->mCreatePending = false;
         mCreateTextureList.pop();
 
         if (imagep->hasGLTexture() && imagep->getDiscardLevel() < imagep->getDesiredDiscardLevel())
         {
+            // NOTE: this may happen if the desired discard reduces while a decode is in progress and does not
+            // necessarily indicate a problem, but if log occurrences excede that of dsiplay_stats: FPS,
+            // something has probably gone wrong.
```
Comment (akleshchev) [inline indra/newview/llviewertexturelist.cpp:1088]: Scaling is new, so might be a good idea to add some kind of debug logging here, that shows Id and by how much image gets scaled.
Or compare to mDebugId then log details.

And texture in a 'scaled' state probably should be represented in texture console somehow.

**Pull request merged**